File: classify_train.jl
# Copyright (c) 2026 Mel Young.
#
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL
# was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
using TextAnalysis: serialize
using SQLite
using DataFrames
using WordTokenizers
using StatsBase
using Serialization
using MLDataUtils
using TextAnalysis
using ProgressMeter
using MLJ
using MLJText
using MLJBase
using Languages
using ThreadsX

CountTransformer = @load CountTransformer pkg=MLJText
MultinomialNBClassifier = @load MultinomialNBClassifier pkg=NaiveBayes

# Trains the classifiers on the spam (AI) and ham (human) datasets

function load_data()
    println("Fetching data from db...")
    db = SQLite.DB("data/panslop.db")
    spam_db = DBInterface.execute(db, "SELECT * FROM full_text ORDER BY RANDOM() LIMIT 2000") |> DataFrame
    ham_db = DBInterface.execute(db, "SELECT * FROM ham_full_text ORDER BY RANDOM() LIMIT 2000") |> DataFrame

    # ham_linux_docs = read("ham/LINUX_DOCS.md", String)
    # println("Loaded.")

    spam = spam_db.text
    ham = ham_db.text
    # push!(ham, ham_linux_docs)

    return spam, ham
end

function train()
    spam, ham = load_data()
    corpus = vcat(spam, ham)
    println("$(length(spam)) spam files")
    println("$(length(ham)) ham files")

    # split test and train set with Julia's cool new MLDataUtils
    # refs:
    # https://discourse.julialang.org/t/simple-tool-for-train-test-split/473/4
    # https://github.com/JuliaML/MLDataUtils.jl
    train_ham, test_ham = splitobs(ham; at=0.8)
    train_spam, test_spam = splitobs(spam; at=0.8)

    # prepare labels on the train set
    train_corpus = vcat(train_ham, train_spam)
    labels = vcat(repeat(["ham"], length(train_ham)), repeat(["spam"], length(train_spam)))

    # based on:
    # https://juliaai.github.io/MLJ.jl/stable/models/MultinomialNBClassifier_NaiveBayes/#MultinomialNBClassifier_NaiveBayes'
    # https://github.com/JuliaAI/MLJText.jl#tf-idf-transformer

    println("Tokenising...")
    tokenised = ThreadsX.map(doc -> TextAnalysis.tokenize(Languages.English(), doc), corpus)

    println("Computing TF-IDF features...")
    mach1 = machine(CountTransformer(), tokenised) |> MLJ.fit!

    # matrix of counts
    X = MLJ.transform(mach1, tokenised)
    y = coerce(labels, OrderedFactor)
    serialize("data/corpus.dat", corpus)
    serialize("data/X.dat", X)
    serialize("data/Y.dat", y)
    serialize("data/mach1.dat", mach1)
    serialize("data/tokenised.dat", tokenised)

    println("Now training...")
    classifier = MultinomialNBClassifier()
    mach2 = machine(classifier, X, y)
    MLJ.fit!(mach2, rows=1:length(train_corpus))
    serialize("data/mach2.dat", mach2)

    println("Done.")
end

train()

File: classify_train.py
# Copyright (c) 2026 Mel Young.
#
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL
# was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
import pickle
import sqlite3
import matplotlib.pyplot as plt

import numpy as np
import pandas as pd
import rich
from IPython import embed
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegressionCV
from sklearn.metrics import accuracy_score, classification_report, log_loss
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

# based on: https://github.com/nadinejackson1/text-classification-naive-bayes/blob/main/main.ipynb


def load_data() -> pd.DataFrame:
    conn = sqlite3.connect("data/panslop.db")
    spam = pd.read_sql_query(
        "SELECT text FROM full_text ORDER BY RANDOM() LIMIT 6000", conn
    )
    ham = pd.read_sql_query(
        "SELECT text FROM ham_full_text ORDER BY RANDOM() LIMIT 6000", conn
    )
    conn.close()

    spam["label"] = "spam"
    ham["label"] = "ham"

    return pd.concat([spam, ham], ignore_index=True)


def train():
    print("Loading data...")
    data: pd.DataFrame = load_data().sample(frac=1)
    print(data)

    X = data["text"]
    y = data["label"]
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    print("Vectorising...")
    vectorizer = CountVectorizer(stop_words="english", max_features=10000)
    X_train_vect = vectorizer.fit_transform(X_train)
    X_test_vect = vectorizer.transform(X_test)

    # based on https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html
    classifiers = {
        "Decision Tree": DecisionTreeClassifier(max_depth=5),
        "Naive Bayes": MultinomialNB(),
        "KNN": KNeighborsClassifier(5),
        "Linear SVM": SVC(kernel="linear", C=0.025, random_state=42),
        "Random Forest": RandomForestClassifier(
            max_depth=5, n_estimators=10, max_features=5, random_state=42
        ),
        # "Logistic Regression": LogisticRegressionCV(
        #     Cs=np.logspace(-6, 6, 101),
        #     cv=10,
        #     l1_ratios=(0,),
        #     scoring="neg_log_loss",
        #     max_iter=1_000,
        #     use_legacy_attributes=False,
        # ),
    }

    for name, classifier in classifiers.items():
        print(f"Fitting classifier {name}...")
        classifier.fit(X_train_vect, y_train)

        print("Predicting...")
        y_pred = classifier.predict(X_test_vect)
        report = classification_report(y_test, y_pred)

        rich.print(
            f"[bold red]Classification report for {name}:[/bold red]\n[lime]{
                report
            }[/lime]"
        )
        print()

        if isinstance(classifier, DecisionTreeClassifier):
            tree.plot_tree(classifier, class_names=True)
            plt.show()

    # print("SPAM most salient:")
    # spam_class_prob_sorted = classifier.feature_log_prob_[0, :].argsort()[::-1]
    # print(np.take(vectorizer.get_feature_names_out(), spam_class_prob_sorted[:10]))
    #
    # print("HAM most salient:")
    # ham_class_prob_sorted = classifier.feature_log_prob_[1, :].argsort()[::-1]
    # print(np.take(vectorizer.get_feature_names_out(), ham_class_prob_sorted[:10]))


if __name__ == "__main__":
    train()
